File: src/rule_engine.py
"""Declarative rule evaluator for the moneyline confidence engine (Python side).

The rule catalog lives in data/rules/moneyline_rules.json -- the single source
of truth shared with the JavaScript evaluator (src/rule_engine.js). This module
owns ONLY the per-language predicate logic (the ``handler`` functions); every
threshold, message, ordering, engine-scope and tier lives in the JSON.

Behavior contract: evaluate_moneyline() must reproduce the exact reason and
adjustment strings, the running-confidence mutations, and the ordering that the
old inline middle block of apply_confidence_downgrade() produced in
src/quality_control.py. Rules fire in ascending ``order`` (which mirrors the
original source line sequence); ``tier`` is descriptive metadata only and does
NOT affect ordering.

Numeric formatting stays in the handlers because JS (.toFixed) and Python
(:.0f) format differently, so it cannot live in the shared JSON.
"""

from __future__ import annotations

import functools
import json
import os
from typing import Any, Callable

from .utils import data_path

# Confidence ladder + helpers are duplicated here (rather than imported from
# quality_control) so the engine has no import cycle back into its host. They
# must stay byte-identical with quality_control._CONFIDENCE_LEVELS etc.
_CONFIDENCE_LEVELS = ("Low", "Medium", "High")


def _rules_path():
    override = os.environ.get("MLB_RULES_FILE")
    if override:
        from pathlib import Path

        return Path(override)
    return data_path("rules/moneyline_rules.json")


@functools.lru_cache(maxsize=1)
def _load_rules_cached(path_str: str) -> dict[str, Any]:
    with open(path_str, "r", encoding="utf-8") as handle:
        return json.load(handle)


def load_moneyline_rules() -> dict[str, Any]:
    """Return the parsed rule catalog (cached per resolved path)."""
    return _load_rules_cached(str(_rules_path()))


def _reset_rules_cache() -> None:
    """Test-only: drop the cache so a different MLB_RULES_FILE can be loaded."""
    _load_rules_cached.cache_clear()


def _normalize_confidence(value: Any) -> str:
    text = str(value or "Low").strip().title()
    return text if text in _CONFIDENCE_LEVELS else "Low"


def _cap_confidence(confidence: str, cap: str) -> str:
    current_index = _CONFIDENCE_LEVELS.index(_normalize_confidence(confidence))
    cap_index = _CONFIDENCE_LEVELS.index(_normalize_confidence(cap))
    return _CONFIDENCE_LEVELS[min(current_index, cap_index)]


def _downgrade(confidence: str) -> str:
    index = _CONFIDENCE_LEVELS.index(_normalize_confidence(confidence))
    return _CONFIDENCE_LEVELS[max(0, index - 1)]


# ---------------------------------------------------------------------------
# Handlers. Each mirrors one check from the original middle block of
# apply_confidence_downgrade() (src/quality_control.py:394-466). A handler
# receives (state, ctx, params) and mutates ``state`` in place:
#   state["no_bet"]      -> bool
#   state["confidence"]  -> str (running confidence, threaded across CAP/ADJUST)
#   state["reasons"]     -> list[str] (NO_BET reason strings)
#   state["adjustments"] -> list[str] (downgrade/cap adjustment strings)
# ctx carries the resolved inputs the host computed (quality_report fields,
# edge, market_type, edge threshold, sharp adjustment classification).
# ---------------------------------------------------------------------------

MISSING = "Missing"
PROJECTED = "Projected"
STALE = "Stale"


def _format_edge_threshold(value: float) -> str:
    return f"{value * 100:.0f}%"


def _probable_pitcher_missing(state, ctx, params, rule) -> None:
    if ctx.get("probable_pitchers") == MISSING:
        state["no_bet"] = True
        state["reasons"].append(rule["message"])


def _opener_consideration(state, ctx, params, rule) -> None:
    if "opener_situation" in (ctx.get("no_bet_considerations") or []):
        state["adjustments"].append(rule["message"])


def _edge_unavailable(state, ctx, params, rule) -> None:
    if ctx.get("edge_value") is None:
        state["no_bet"] = True
        state["reasons"].append(rule["message"])


def _yrfi_edge_floor(state, ctx, params, rule) -> None:
    edge_value = ctx.get("edge_value")
    threshold = params.get("threshold", 0.06)
    if edge_value is not None and ctx.get("market_type") == "yrfi" and abs(edge_value) < threshold:
        state["no_bet"] = True
        state["reasons"].append(rule["message"])


def _edge_floor(state, ctx, params, rule) -> None:
    edge_value = ctx.get("edge_value")
    threshold = ctx.get("edge_threshold")
    if (
        edge_value is not None
        and ctx.get("market_type") != "yrfi"
        and abs(edge_value) < threshold
    ):
        state["no_bet"] = True
        state["reasons"].append(
            rule["message"].format(threshold_pct=_format_edge_threshold(threshold))
        )


def _data_quality_floor(state, ctx, params, rule) -> None:
    if int(ctx.get("score", 0)) < params.get("min_score", 60):
        state["no_bet"] = True
        state["reasons"].append(rule["message"])


def _sharp_downgrade(state, ctx, params, rule) -> None:
    sharp_adj = ctx.get("sharp_adj", "no_change")
    if sharp_adj == "downgrade_two":
        state["confidence"] = _downgrade(_downgrade(state["confidence"]))
        state["adjustments"].append(
            "sharp money strongly against pick: confidence downgraded x2"
        )
    elif sharp_adj == "downgrade_one":
        state["confidence"] = _downgrade(state["confidence"])
        state["adjustments"].append("sharp money against pick: confidence downgraded")


def _odds_stale_downgrade(state, ctx, params, rule) -> None:
    if ctx.get("odds") == STALE:
        state["confidence"] = _downgrade(state["confidence"])
        state["adjustments"].append(rule["message"])


def _weather_stale_downgrade(state, ctx, params, rule) -> None:
    if ctx.get("weather") == STALE and ctx.get("weather_outdoor"):
        state["confidence"] = _downgrade(state["confidence"])
        state["adjustments"].append(rule["message"])


def _lineup_cap(state, ctx, params, rule) -> None:
    if ctx.get("lineup") in {PROJECTED, MISSING}:
        new_confidence = _cap_confidence(state["confidence"], rule["cap"])
        if new_confidence != state["confidence"]:
            state["adjustments"].append(rule["message"])
        state["confidence"] = new_confidence


def _pitcher_projected_cap(state, ctx, params, rule) -> None:
    if ctx.get("probable_pitchers") == PROJECTED:
        new_confidence = _cap_confidence(state["confidence"], rule["cap"])
        if new_confidence != state["confidence"]:
            state["adjustments"].append(rule["message"])
        state["confidence"] = new_confidence


def _score_band_cap(state, ctx, params, rule) -> None:
    score = int(ctx.get("score", 0))
    for band in params.get("bands", []):
        if band["min"] <= score < band["max"]:
            if band.get("requires_calibration"):
                # 85+ band: only fires for High confidence lacking calibration
                # support, and unconditionally sets (not caps) to the band cap.
                if state["confidence"] == "High" and not ctx.get(
                    "calibration_supports_high"
                ):
                    state["confidence"] = band["cap"]
                    state["adjustments"].append(band["message"])
            else:
                new_confidence = _cap_confidence(state["confidence"], band["cap"])
                if new_confidence != state["confidence"]:
                    state["adjustments"].append(band["message"])
                state["confidence"] = new_confidence
            # elif chain: at most one band applies (ranges are disjoint).
            return


PY_HANDLERS: dict[str, Callable[..., None]] = {
    "probablePitcherMissing": _probable_pitcher_missing,
    "openerConsideration": _opener_consideration,
    "edgeUnavailable": _edge_unavailable,
    "yrfiEdgeFloor": _yrfi_edge_floor,
    "edgeFloor": _edge_floor,
    "dataQualityFloor": _data_quality_floor,
    "sharpDowngrade": _sharp_downgrade,
    "oddsStaleDowngrade": _odds_stale_downgrade,
    "weatherStaleDowngrade": _weather_stale_downgrade,
    "lineupCap": _lineup_cap,
    "pitcherProjectedCap": _pitcher_projected_cap,
    "scoreBandCap": _score_band_cap,
}


def evaluate_moneyline(ctx: dict[str, Any]) -> dict[str, Any]:
    """Evaluate all py-scoped rules against ``ctx`` in ascending ``order``.

    ``ctx`` must supply: confidence, probable_pitchers, no_bet_considerations,
    edge_value, market_type, edge_threshold, score, sharp_adj, odds, weather,
    weather_outdoor, lineup, calibration_supports_high.

    Returns {no_bet, reasons, adjustments, confidence}. The running confidence
    is threaded across CAP/ADJUST handlers so they compound in source order,
    exactly as the old inline middle block did.
    """
    rules = sorted(
        (rule for rule in load_moneyline_rules()["rules"] if "py" in rule["engines"]),
        key=lambda rule: rule["order"],
    )

    state: dict[str, Any] = {
        "no_bet": False,
        "confidence": _normalize_confidence(ctx.get("confidence")),
        "reasons": [],
        "adjustments": [],
    }

    for rule in rules:
        handler = PY_HANDLERS.get(rule["handler"])
        if handler is None:
            raise ValueError(
                f"No Python handler registered for rule {rule['id']} "
                f"(handler: {rule['handler']})"
            )
        handler(state, ctx, rule.get("params") or {}, rule)

    return state

File: src/rule_engine.js
// Declarative rule evaluator for the moneyline value engine (JS side).
//
// The rule catalog lives in data/rules/moneyline_rules.json — the single source
// of truth shared with the Python evaluator (src/rule_engine.py). This module
// owns ONLY the per-language predicate logic (the `handler` functions); every
// threshold, message, ordering, engine-scope and tier lives in the JSON.
//
// Behavior contract: evaluateMoneyline() must reproduce the exact reason strings
// and ordering that the old inline valueSafetyReasons() produced in src/mlb.js.
// Rules fire in ascending `order` (which mirrors the original source line
// sequence); `tier` is descriptive metadata only and does NOT affect ordering.
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { toNumber } from './utils.js';

let _rulesCache = null;

function rulesPath() {
  return process.env.MLB_RULES_FILE || fileURLToPath(new URL('../data/rules/moneyline_rules.json', import.meta.url));
}

export function loadMoneylineRules() {
  if (_rulesCache) return _rulesCache;
  _rulesCache = JSON.parse(readFileSync(rulesPath(), 'utf8'));
  return _rulesCache;
}

// Test-only: drop the cache so a different MLB_RULES_FILE can be loaded.
export function _resetRulesCache() {
  _rulesCache = null;
}

function renderMessage(template, tokens, override) {
  if (override !== undefined && override !== null) return override;
  if (!tokens) return template;
  return template.replace(/\{(\w+)\}/g, (_, key) => String(tokens[key]));
}

function absEdge(item, key) {
  return Math.abs(toNumber(item.modelBreakdown?.[key], 0));
}

// Each handler mirrors one check from the original valueSafetyReasons()
// (src/mlb.js). It returns { fired } plus either `tokens` (filled into the
// rule's message template) or `override` (a pre-built string used verbatim).
export const JS_HANDLERS = {
  // src/mlb.js:529-530 — stale/absent odds timestamp. Host pre-computes the
  // freshness string (config-driven max age) and passes it via ctx.
  staleOdds(ctx) {
    const reason = ctx.oddsFreshnessReason;
    return reason ? { fired: true, override: reason } : { fired: false };
  },

  // src/mlb.js:533-535 — configurable value-edge floor.
  edgeFloor(ctx) {
    const { option, edgeThreshold } = ctx;
    if (option.edge < edgeThreshold) {
      return {
        fired: true,
        tokens: {
          sign: option.edge >= 0 ? '+' : '',
          edge: option.edge.toFixed(1),
          threshold: edgeThreshold.toFixed(1)
        }
      };
    }
    return { fired: false };
  },

  // src/mlb.js:541-545 — picked-team season win% quality floor.
  teamQuality(ctx, params) {
    const floor = toNumber(params.min_win_pct, 0.52);
    if (ctx.pickedTeamWinPct < floor) {
      return {
        fired: true,
        tokens: {
          pct: (ctx.pickedTeamWinPct * 100).toFixed(0),
          floor: (floor * 100).toFixed(0)
        }
      };
    }
    return { fired: false };
  },

  // src/mlb.js:551-553 — away underdog plus-money ceiling.
  awayUnderdog(ctx, params) {
    const { option } = ctx;
    const max = toNumber(params.max_odds, 115);
    if (option.side === 'away' && Number(option.odds) > max) {
      return { fired: true, tokens: { odds: option.odds, max } };
    }
    return { fired: false };
  },

  // src/mlb.js:559-562 — pick must be the model's favored side.
  marketAgreement(ctx) {
    return { fired: ctx.option.side !== ctx.modelFavoredSide };
  },

  // src/mlb.js:566-568 — calibrated conviction floor.
  convictionFloor(ctx, params) {
    const floor = toNumber(params.min_probability, 52);
    const prob = toNumber(ctx.option.modelProbability, 0);
    if (prob < floor) {
      return { fired: true, tokens: { prob: prob.toFixed(1), floor: String(floor) } };
    }
    return { fired: false };
  },

  // src/mlb.js:575-578 — sharp money moved against the pick.
  sharpContra(ctx, params) {
    const sharp = ctx.item.modelBreakdown?.sharpMoney;
    const min = toNumber(params.min_magnitude, 10);
    if (sharp && sharp.direction === 'against_model' && sharp.magnitude >= min) {
      return { fired: true, tokens: { magnitude: sharp.magnitude } };
    }
    return { fired: false };
  },

  // src/mlb.js:580-584 — season record / H2H dominates today's matchup edge.
  recordDominance(ctx, params) {
    const matchupEdge = absEdge(ctx.item, 'matchupEdge');
    const recordContextEdge = absEdge(ctx.item, 'recordContextEdge');
    const mult = toNumber(params.record_multiplier, 1.35);
    const maxMatchup = toNumber(params.max_matchup_edge, 0.2);
    const fired = ctx.item.modelBreakdown?.recordDominated
      || (recordContextEdge > matchupEdge * mult && matchupEdge < maxMatchup);
    return { fired: Boolean(fired) };
  },

  // src/mlb.js:586-596 — evolution-approved record-bias guardrail (params from
  // the active rule; only fires when that rule is present).
  auditRecordBias(ctx, params) {
    const rule = ctx.getEvolutionRule(ctx.evolutionControls, params.evolution_rule_key);
    if (!rule) return { fired: false };
    const rp = rule.parameters || {};
    const matchupEdge = absEdge(ctx.item, 'matchupEdge');
    const recordContextEdge = absEdge(ctx.item, 'recordContextEdge');
    const mult = toNumber(rp.record_context_multiplier, toNumber(params.default_record_context_multiplier, 1.25));
    const maxMatchup = toNumber(rp.max_matchup_edge, toNumber(params.default_max_matchup_edge, 0.18));
    const recordDominatedThin = ctx.item.modelBreakdown?.recordDominated && matchupEdge < maxMatchup;
    const recordContextDominates = recordContextEdge > matchupEdge * mult && matchupEdge < maxMatchup;
    return { fired: Boolean(recordDominatedThin || recordContextDominates) };
  },

  // src/mlb.js:598-600 — today's matchup edge is thin and value edge sub-strong.
  thinMatchup(ctx, params) {
    const matchupEdge = absEdge(ctx.item, 'matchupEdge');
    const maxMatchup = toNumber(params.max_matchup_edge, 0.08);
    const strong = toNumber(params.strong_edge, 4.0);
    return { fired: matchupEdge < maxMatchup && ctx.option.edge < strong };
  },

  // src/mlb.js:602-614 — fewer than N model factors agree with the pick.
  factorAgreement(ctx, params) {
    const breakdown = ctx.item.modelBreakdown || {};
    const pickDir = ctx.option.side === 'home' ? 1 : -1;
    const threshold = toNumber(params.agree_threshold, 0.02);
    const components = [
      toNumber(breakdown.matchupEdge, 0),
      toNumber(breakdown.starterEdge, 0),
      toNumber(breakdown.offenseEdge, 0),
      toNumber(breakdown.bullpenEdge, 0),
      toNumber(breakdown.lineupEdge, 0)
    ];
    const agreeing = components.filter((c) => c * pickDir > threshold).length;
    const minAgree = toNumber(params.min_agreeing, 3);
    const strong = toNumber(params.strong_edge, 4.0);
    return { fired: agreeing < minAgree && ctx.option.edge < strong };
  },

  // src/mlb.js:616-632 — evolution-approved weak-edge guardrail.
  auditWeakEdge(ctx, params) {
    const rule = ctx.getEvolutionRule(ctx.evolutionControls, params.evolution_rule_key);
    if (!rule) return { fired: false };
    const rp = rule.parameters || {};
    const matchupEdge = absEdge(ctx.item, 'matchupEdge');
    const maxValueEdge = toNumber(rp.max_value_edge, toNumber(params.default_max_value_edge, 1.0));
    const maxProbEdge = toNumber(rp.max_probability_edge, toNumber(params.default_max_probability_edge, 3.0));
    const maxMatchup = toNumber(rp.max_matchup_edge, toNumber(params.default_max_matchup_edge, 0.05));
    const probEdge = Math.abs(toNumber(ctx.option.modelProbability, 50) - 50);
    const fired = (ctx.option.edge < maxValueEdge || probEdge < maxProbEdge) && matchupEdge < maxMatchup;
    return { fired: Boolean(fired) };
  },

  // src/mlb.js:634-638 — an incomplete lineup with sub-strong value edge.
  lineupIncomplete(ctx, params) {
    const minCount = toNumber(params.min_count, 9);
    const strong = toNumber(params.strong_edge, 4.0);
    const lineups = [ctx.item.lineups?.away, ctx.item.lineups?.home].filter(Boolean);
    const incomplete = lineups.some((l) => !l.confirmed || toNumber(l.count, 0) < minCount);
    return { fired: incomplete && ctx.option.edge < strong };
  },

  // src/mlb.js:640-646 — opener/bulk pitcher with medium+ confidence.
  openerRisk(ctx, params) {
    const confidences = (params.confidences || ['high', 'medium']).map((c) => String(c).toLowerCase());
    const openerTeams = [ctx.item.away, ctx.item.home].filter((t) => t?.openerSituation?.isOpener);
    const risky = openerTeams.some((t) => confidences.includes(String(t.openerSituation?.confidence || '').toLowerCase()));
    return { fired: risky };
  },

  // src/mlb.js:648-651 — a probable pitcher is not yet posted.
  noProbablePitcher(ctx, params) {
    const strong = toNumber(params.strong_edge, 4.0);
    const missing = !ctx.item.away?.starter || !ctx.item.home?.starter;
    return { fired: missing && ctx.option.edge < strong };
  }
};

// Evaluate all JS-scoped rules against ctx and return the deduped, ordered list
// of fired reason strings. Mirrors the old valueSafetyReasons() return value.
// Does NOT short-circuit — collects every fired reason (the NO BET/VALUE label
// is applied by the host in applyMoneylineValueMarket).
export function evaluateMoneyline(ctx) {
  const rules = loadMoneylineRules().rules
    .filter((rule) => rule.engines.includes('js'))
    .sort((a, b) => a.order - b.order);

  const reasons = [];
  for (const rule of rules) {
    const handler = JS_HANDLERS[rule.handler];
    if (!handler) throw new Error(`No JS handler registered for rule ${rule.id} (handler: ${rule.handler})`);
    const result = handler(ctx, rule.params || {});
    if (result?.fired) reasons.push(renderMessage(rule.message, result.tokens, result.override));
  }
  return [...new Set(reasons)];
}
